model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java

```diff
@@ -537,40 +535,43 @@ public void removeLocalUserSessions(String realmId, boolean offline) {
 
         final AtomicInteger userSessionsSize = new AtomicInteger();
 
-        localCacheStoreIgnore
+        removeEntriesByRealm(realmId, localCacheStoreIgnore, userSessionsSize, localCache, localClientSessionCache);
+
+        // iterate over the remote entries as well to ensure they are purged
+        removeEntriesByRealm(realmId, getCache(offline), userSessionsSize, getCache(offline), getClientSessionCache(offline));
```

Comment:
The previous line removes all sessions from the local caches for the realm. This line attempts to remove also remote sessions. 

Would it make sense to use `RemoteCache` directly for remote store removal instead of using `getCache()`. Also, should we add `SKIP_LISTENER_NOTIFICATION` flag to avoid processing of removals by each Keycloak node again? We know the local caches are already clean of sessions for the realm so we don't need to distribute these events. 

If I remember correctly `SKIP_LISTENER_NOTIFICATION` still sends the notification to the other site in case of CrossDC setups but it would still make the number of sent events lower.

Reply:
I've pushed one more commit with the changes you suggested. As the local caches and the remote caches have different interfaces, I needed to duplicate the method to make it work. Please have a look if it now looks better to you.